client/login.jsx

```diff
@@ -22,49 +26,88 @@ var LoginForm = React.createClass({
         return true;
     },
-    _handleError(validationState) {
-        this.setState({
-            validationState: validationState,
-            errorStr: i18n.t('usernameSet.errors.' + validationState)
-        });
+    _validatePassword(password) {
+        if (password == '') {
+            return 'empty';
+        }
+
+        return true;
+    },
+    _handleError(validationStateUsername, validationStatePassword) {
+        if (validationStateUsername !== true) {
+            this.setState({
+                errorStr: i18n.t('usernameSet.errors.' + validationStateUsername)
+            });
+        }
+        else if (validationStatePassword !== true) {
+            this.setState({
+                errorStr: i18n.t('passwordSet.errors.' + validationStatePassword)
+            });
+        }
     },
     _onUsernameChanged(username) {
-        var validationState = this._validate(username);
+        var validationStateUsername = this._validateUsername(username);
+        this.setState({validationStateUsername: validationStateUsername});
 
         this.setState({username});
-        if (!this.state.changed) {
+        if (!this.state.usernameChanged) {
             this.setState({
-                changed: true
+                usernameChanged: true
             });
         }
 
-        this._handleError(validationState);
+        this._handleError(validationStateUsername, this.state.validationStatePassword);
     },
-    onError(error) {
-        this._handleError(error);
+    _onPasswordChanged(password) {
+        var validationStatePassword = this._validatePassword(password);
+        this.setState({validationStatePassword: validationStatePassword});
+
+        this.setState({password});
+        if (!this.state.passwordChanged) {
+            this.setState({
+                passwordChanged: true
+            });
+        }
+
+        this._handleError(this.state.validationStateUsername, validationStatePassword);
+    },
+    onError(errorUsername, errorPassword) {
+        this._handleError(errorUsername, errorPassword);
     },
     onSuccess() {
         $(React.findDOMNode(this.refs.usernameSetModal)).modal('hide');
     },
-    handleChange(event) {
+    handleChangeUsername(event) {
         this._onUsernameChanged(event.target.value);
     },
+    handleChangePassword(event) {
+        this._onPasswordChanged(event.target.value);
+    },
     handleSubmit(event) {
         event.preventDefault();
 
-        if (!this.state.changed) {
+        if (!this.state.usernameChanged) {
             this._onUsernameChanged('');
             return;
         }
 
-        if (this.state.validationState !== true) {
+        if (this.state.type == 'passwordLogin' && !this.state.passwordChanged) {
+            this._onPasswordChanged('');
+            return;
+        }
+
+        if ((this.state.validationStateUsername !== true) ||
+           (this.state.type == 'passwordLogin' && this.state.validationStatePassword !== true)) {
```

Comment:
For readability, this is better split into a separate if: First, check if the username validation failed and return. Next, check if the password validation failed and return:

```
if (this.state.validationStateUsername !== true) {
    return;
}
if (this.state.type == 'passwordLogin') {
    if (this.satte.validationStatePassword !== true) {
        return;
    }
}
```